### ISSUE TITLE
output in the custom loss function is different from one I obtain with predict.

### ISSUE THREAD

**Issue by stamas02** [fchollet/keras#4681]
Ok so I have been trying to figure this out for days now and I think there is some problem with the custom loss function. So I have the following code:

```
self._model.compile(loss=self.custom_loss, optimizer="sgd")
out = self._model.predict(x)
sum1 = np.sum(out)
self._model.fit(x=x,y=y,batch_size=len(x),nb_epoch=1,shuffle=False)

def custom_loss(self, y_true, y_pred):
        return K.sum(y_pred)
```

At this point I would expect the loss and sum1 to be the same but they are very different. Note that this code does not have any meaning but my loss calculation was incorrect and wanted to test what my loss function receives. I do not have any dropout layers.

**Comment by patyork** [comment]
Weight and activity regularizes contribute to the loss as well.

-----Original Message-----

**Comment by stamas02** [comment]
Yes but they would contribute to the output in the same way. Anyway I do not use either of these

**Comment by patyork** [comment]
Regularizers and Dropout are a component in the `fit` loss.

Anyways, the loss from `fit` and `evaluate` are *averages* across the elements in the batch(es) that were seen; the last dimension I believe, depending on tf/th ordering.

```python
out = model.predict(x)
out_loss = np.mean(np.sum(out, axis=-1)) / x.shape[-1]
np.allclose(out_loss, model.evaluate(x, y))
>> True
```

**Comment by stamas02** [comment]
I only feed 1 batch with 1 image to the cnn in the fit and in the predict funtion as well (same image) and it outputs a 128dimensional feature vector. I define my own loss function as I wrote. So out is a 128 dimesnional feature vector and its shape is (1,128). Then the sum basically returns a scaler so the mean doesnt actually do anithing and y.shape[-1] is 1 so it doesnt change the value either

**Comment by patyork** [comment]
```python
a = np.ones((1,128))
a.shape[-1]
>> 128
```

**Comment by stamas02** [comment]
Ok so in your code
`out_loss = np.mean(np.sum(out, axis=-1)) / x.shape[-1]`

where x is the inout to the net so it is my image with shape (width,height,channels) and shape[-1] is 3 with rgb. what does the channel number has to do with the loss?

**Comment by stamas02** [comment]
Ok so I reduced the output dimensionality to 1 and the two numbers I still not the same. I then removed all the batch normalization layers and now it suddenly works. So I will need to have a deeper look how batch normalization is implemented but I cannot imagine how it would be possibly right. So let me once again try and clearly describe what is happening here.

```
self._model.compile(loss=my_loss, optimizer="sgd")
imgs = np.array([cv2.imread("some image") # shape = (1,100,100,3)])
out = self._model.predict(imgs, batch_size = len(imgs))  # out shape = (1,1)
 self._model.fit(x=imgs,y=y,batch_size=len(imgs),nb_epoch=1,shuffle=False)

def my_loss(self, y_true, y_pred):
     return K.flatten(y_pred)
```

Then the loss and out has to be the same and they are the same without batch normalization and they are fundamentally different with batch normalization.

**Comment by patyork** [comment]
The batch normalization can be read about more[ in the doc](https://keras.io/layers/normalization/).

The behavior is significantly different depending on the type ("mode") the layer is set in. In default mode 0, the output will be different depending on whether you call `fit`, `evaluate` or `predict`. In mode 2, those calls should return the same data, batch-by-batch, excluding after `fit` is called and the weights change.

This is the note about mode 0, using different values in train/test to normalize data:
>During training we use per-batch statistics to normalize the data, and during testing we use running averages computed during the training phase.

And mode 2:
>like mode 0, but using per-batch statistics to normalize the data during both testing and training.

The modes essentially control over which axis the input is normalized on, and whether it is normalized on a per-batch basis, or if a more complicated method of remembering prior normalizations and applying them.

I'm guessing your BatchNorm is set to the default mode 0 - if you change it to mode 2 you will probably see similar values.

**Comment by stamas02** [comment]
Thanks for the quick reply! I will have a look. I will close this now
